### PR TITLE
Fixes #8433 - non-apostrophed ('...') url with multiple params support in cURL import

### PR DESCRIPTION
**Pull request by SahilChoudhary22**
- Fixes https://github.com/postmanlabs/postman-app-support/issues/8843
- Added few more test cases to support regex changes.

### CODE REVIEW

File: test/conversion.test.js [hunk 1]
```diff
@@ -594,4 +594,115 @@ describe('Curl converter should', function() {
...
+        -H "pid: bdb5748e" \\
+        -H "sid: 5096fd6e" \\
+        -H "x-features: FEATURE1_ENABLED" \\
+        -H "Accept-Language: tr-TR" \\
+        -H "Content-Type: application/json" \\
+        https://test.com/test/foo?bar=1&baz=2`
+      }, function (err, result) {
+        expect(result.result).to.equal(true);
+        expect(result.output.length).to.equal(1);
+        expect(result.output[0].type).to.equal('request');
```
Comment (umeshp7) [inline test/conversion.test.js:622]: The test has to expect the right information you are expecting inside the request as well. For eg. the url in this case.

File: assets/shell-quote.js [hunk 2]
```diff
@@ -25,6 +25,7 @@ var META = '|&;()<> \\t';
 var BAREWORD = '(\\\\[\'"' + META + ']|[^\\s\'"' + META + '])+';
```
Comment (umeshp7) [inline assets/shell-quote.js:25]: @SahilChoudhary22 This is an external asset that we have added to the module. We will have to be completely sure before we make any changes to it.

Comment (umeshp7) [inline assets/shell-quote.js:25]: @SahilChoudhary22 This is an external asset that we have added to the module. We will have to be completely sure before we make any changes to it.

File: test/conversion.test.js [hunk 3]
```diff
@@ -594,4 +594,115 @@ describe('Curl converter should', function() {
...
+        -H "x-application-id: 5" \\
+        -H "Build: 4.6.4.459" \\
+        -H "Platform: Android" \\
+        -H "Gender: M" \\
+        -H "SearchSegment: 98" \\
+        -H "OSVersion: 10" \\
+        -H "DeviceId: b147ee89" \\
+        -H "pid: bdb5748e" \\
+        -H "sid: 5096fd6e" \\
+        -H "x-features: FEATURE1_ENABLED" \\
```
Comment (umeshp7) [inline test/conversion.test.js:615]: We can remove some of the unwanted headers etc. here and use the smallest version which reproduces the issue.

Comment (umeshp7) [inline test/conversion.test.js:615]: We can remove some of the unwanted headers etc. here and use the smallest version which reproduces the issue.

File: assets/shell-quote.js [hunk 4]
```diff
@@ -25,6 +25,7 @@ var META = '|&;()<> \\t';
...
 var SINGLE_QUOTE = '"((\\\\"|[^"])*?)"';
 var DOUBLE_QUOTE = '\'((\\\\\'|[^\'])*?)\'';
+var HTTP_URL_MATCHER  = "http[^ ]+";
```
Comment (umeshp7) [inline assets/shell-quote.js:28]: @SahilChoudhary22 can we check the shell-quote npm module, if there are any open issues around this or if there is any handling already for this?

Comment (umeshp7) [inline assets/shell-quote.js:28]: @SahilChoudhary22 can we check the shell-quote npm module, if there are any open issues around this or if there is any handling already for this?

Comment (umeshp7) [reply assets/shell-quote.js:28]: The url could simply be www.google.com?foo=bar&foo=baz or google.com?foo=bar&foo=baz and these cases aren't covered with this.

**Review by umeshp7 — CHANGES_REQUESTED** [review]

**Comment by SahilChoudhary22** [comment]
- Rolled back the changes in `assets/shell-quote.js` to avoid modification in an external library. 
- Made changes in `src/lib.js` and added cURL string preprocessing to handle incorrectly formed input.
- Compacted the test cases, removed unnecessary headers. Added some more assertions in each test for proper testing.

File: test/conversion.test.js [hunk 5]
```diff
@@ -594,4 +594,177 @@ describe('Curl converter should', function() {
...
+      });
+    });
+    it('in case where there is multiple params with & in between in the url (param after route)', function(done) {
+      convert({
+        type: 'string',
+        data: `curl -X GET \\
+        -H "User-Agent: Dalvik/2.1.0 (Linux; U; Android 10; SM-A705FN Build/QP1A.190711.020) Test/4.6.4.459" \\
+        -H "Authorization: bearer XXX" \\
+        -H "Content-Type: application/json" \\
+        "test.com/?bar=1&baz=2"`
```
Comment (umeshp7) [inline test/conversion.test.js:757]: let's add a test with more than 2 query params.

File: src/lib.js [hunk 6]
```diff
@@ -466,7 +466,11 @@ var program,
         this.initialize();
         this.requestUrl = '';
 
-        var sanitizedArgs = this.sanitizeArgs(curlString),
+        // [Github #8843] - RegEx to fix malformed cURLs with unquoted multi-param URLs
+        const multiParamUrlRegEx = /[^ ]+\.[^ ]+&[^ ]+/gm;
+
+        var cleanedCurlString = curlString.replace(multiParamUrlRegEx, `'${curlString.match(multiParamUrlRegEx)}'`),
```
Comment (umeshp7) [inline src/lib.js:472]: Will single quotes always work?

File: src/lib.js [hunk 7]
```diff
@@ -466,7 +466,11 @@ var program,
         this.initialize();
         this.requestUrl = '';
 
-        var sanitizedArgs = this.sanitizeArgs(curlString),
+        // [Github #8843] - RegEx to fix malformed cURLs with unquoted multi-param URLs
+        const multiParamUrlRegEx = /[^ ]+\.[^ ]+&[^ ]+/gm;
```
Comment (umeshp7) [inline src/lib.js:470]: What will happen if we have a URL like this > 
`https://postman-echo.com/get?foo1=bar1&` 

@SahilChoudhary22

**Comment by umeshp7** [comment]
We should add additional handling for trailing `&` in the URL which is not accounted for in the current regex.

**Comment by umeshp7** [comment]
We should safe check at `arg.startsWith('$')`. And throw appropriate errors where necessary. 

If this `&` is still encountered after the handling like:`curl https://postman-echo.com/post -d foo1=bar1&foo2=bar2` these are clearly invalid. Then we should respond with an error saying "Error parsing curl: <message>"

File: test/conversion.test.js [hunk 8]
```diff
@@ -594,4 +594,269 @@ describe('Curl converter should', function() {
...
+    it('in case where there is multiple params with & in between in the url (https)', function(done) {
+      convert({
+        type: 'string',
+        data: `curl -X GET \\
+        -H "User-Agent: Dalvik/2.1.0 (SM-A705FN Build/QP1A.190711.020)" \\
+        -H "Authorization: bearer XXX" \\
+        -H "Build: 4.6.4.459" \\
+        -H "Platform: Android" \\
+        -H "Accept-Language: tr-TR" \\
+        -H "Content-Type: application/json" \\
```
Comment (umeshp7) [inline test/conversion.test.js:608]: Do we need all these headers?

**Review by umeshp7 — APPROVED** [review]

**Comment by umeshp7** [comment]
@SahilChoudhary22 Looks good to me. 
Let's merge this and we can test this one beta.

**Pull request merged**